R/Quosure.R

```diff
@@ -0,0 +1,307 @@
+#' Reproducible class with environment and code.
+#'
+#' @slot code (`character`) to reproduce the environment
+#' @slot env (`environment`) environment which content was generated by the evaluation
+#' of the `code` slot.
+setClass(
+  "Quosure",
+  representation(code = "character", env = "environment"),
+  prototype(code = character(0), env = new.env(parent = parent.env(.GlobalEnv)))
+)
+
+#' Initialize `Quosure` object
+#'
+#' Initialize `Quosure` object with `code` and `env`. In order to have `Quosure` reproducible
+#' one needs to initialize with `env` which can be reproduced by the `code`. Alternatively, one
+#' can create an empty `Quosure` and evaluate the expressions in this object.
+#' @name new_quosure
+#'
+#' @param code (`character(1)` or `language`) code to evaluate. Also acceptps and stores comments
+#' @param env (`environment`) name of the code element.
+#'
+#' @export
+setGeneric("new_quosure", function(code = character(0), env = parent.env(.GlobalEnv)) {
+  standardGeneric("new_quosure")
+})
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure", signature(code = "missing", env = "missing"),
+  function(code, env) {
+    methods::new("Quosure")
+  }
+)
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure", signature(code = "character", env = "environment"),
+  function(code = character(0), env = parent.env(.GlobalEnv)) {
+    checkmate::check_class(env, "environment")
+    checkmate::check_string(code)
+
+    lockEnvironment(env)
+
+    if (is.null(names(code))) {
+      names(code) <- "code"
+    }
+    methods::new("Quosure", code = code, env = env)
+  }
+)
+
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure", signature(code = "character", env = "list"),
+  function(code, env) {
+    env <- if (checkmate::check_list(env, "reactive")) {
+      lapply(env, function(x) {
+        if (inherits(x, "reactive")) {
+          x()
+        } else {
+          x
+        }
+      })
+    }
+
+    new_quosure(code = code, env = list2env(env))
+  }
+)
+
+
+setGeneric("get_code", function(object, deparse = FALSE) {
+  standardGeneric("get_code")
+})
+
+setGeneric("join", function(object, object2, overwrite = FALSE) {
+  standardGeneric("join")
+})
+
+#' Evaluate the code in the `Quosure` environment
+#'
+#' Given code is evaluated in the `Quosure` environment and appended to the `code` slot. This means
+#' that state of the environment is always a result of the stored code (if `Quosure` was initialized)
+#' with reproducible code.
+#'
+#' @name eval_code
+#'
+#' @param object (`Quosure`)
+#' @param code (`character(1)` or `language`) code to evaluate. Also acceptps and stores comments
+#' @param name (`character(1)`) name of the code element.
```

Comment:
what's the point of this argument?

Reply:
to name the elements of the code (to name a vector). Might be needed to do something like `get_eval_details_srv ` . It will be possible also to keep more informations like warnings, errors and notes also from the evaluations

Reply:
Do we have to name elements of the code vector? Maybe YAGNI and start simple?